lib/build-cron.js

```
@@ -22,8 +23,11 @@ module.exports = function buildCronJobs(cronConfigsDir, collectionName, logClean
 
   if (!cronConfigsDir || !collectionName) {
     const err = new VError('Error: \'cronConfigsDir\' and \'collectionName\' are required params.');
-    raygun.send(err);
-    throw err;
+    return Promise.resolve(raygun.send(err))
```

Does `raygun.send(err)` return a promise?

---

That is correct

---

Okay so I don't know if you need Bluebird then. Bc you could just go `raygun.send(err).then(...`

---

See, that's what I thought as well. But for whatever reason, the default Promises don't seem to handle errors thrown within the Promise chain the same way.  I had originally tried doing the same thing, without bringing Bluebird in, but the errors were just swallowed

---

Do you mean if the `raygun.send(err)` call throws an error? Have you tried doing:

`raygun.send(err).then(...).catch(...)`

---

What I mean is, when using `raygun.send(err)`, the `throw err` line within the `.then()` gets swallowed.  Everything continues on as normal, no error is ever seen, as if it's not even there.  When using Bluebird via `Promise.resolve(raygun.send(err))`, everything works as expected, and the error gets propagated up the chain

---

Ahh okay....so final question, why are we doing promises at all?

Why not just: 
```
if (!cronConfigsDir || !collectionName) {
    const err = new VError('Error: \'cronConfigsDir\' and \'collectionName\' are required params.');
    raygun.send(err);
    log.info('raygun error sent');
    throw err;
  }
```

---

That's what I originally had, and I figured that would work.  But because that of code is run synchronously, the `throw err` ends the node process before the `raygun.send(err)` has resolved, resulting in the error not showing up in Raygun

---

Okay I gothcha now. So there are some places where you are doing a `.then(....` and others where you are doing a `asCallback(....`. Why the difference?

---

So, that's where I got really bogged down in all this.  So if you look in `build-cron.js`, all the places where we use `.asCallback()` are within the callback function of `cmaster.loadJobs()`.   `raygun.send()` returning a promise was an issue, and I was once again having problem getting the darn error to propagate after being thrown.  The only way around it that I found was to attach the `throw err` as a callback to `raygun.send()` using the `.asCallback()` function of Bluebird

---

Hmm alright man. Have you deployed this and tested it out?

---

I've tested it as much as I can without actually having the changed merged. The plan is: After this PR is merged, tag it and create a new release, then create a PR to update the version in one of the SS services, and let the cron jobs run overnight in the dev environment

---

Yeah mixing the two can be tricky. Honestly @MikeyBurkman could take one look at it and probably know better. Let me reach out to him and see.

---

So the reason this function threw errors is because we basically want the app to crash. That's it, in a nutshell. 

Generally, you call the build function in your app initialization, and if the initialization fails for any reason, we just simply crash. Fail-fast behavior. There's no real sense in propagating the error up in a promise, because there's arguably no real way of recovering from it -- your app is in a bad, probably-unrecoverable state.

(If the need ever arises where we DO want to allow initialization to fail, then maybe we can add an opt-in parameter or callback or something. But I don't anticipate any of our projects wanting that.)